File: src/lib/azureAuth.ts
import { invoke } from "@tauri-apps/api/core";

// ── Entra ID（App Registration / client credentials）認證 ──────
//
// Azure / Microsoft Foundry 資源常設 disableLocalAuth=true（僅 Entra），
// 因此需以 service principal 換取 bearer token 呼叫資料面 REST。
// token 約 60–90 分到期且無 refresh token，過期後重打即可，這裡做記憶體快取。

export interface AzureEntraCredentials {
  tenantId: string;
  clientId: string;
  clientSecret: string;
}

// Entra token scope（受眾）依「呼叫的 API 路徑」決定，而非 endpoint host：
//  - v1 推論路徑（/openai/v1/...，本專案 chat 整理）→ ai.azure.com 受眾
//  - 傳統 deployments 資料面（/openai/deployments/.../audio/transcriptions?api-version=，
//    本專案 Whisper 轉錄）→ cognitiveservices.azure.com 受眾
// 同一個 Foundry / Azure OpenAI 資源上，這兩條路徑各自要求不同受眾，
// 因此不能用 host 推導（否則其中一條路徑會拿到錯受眾的 token 而 401）。
export const AZURE_SCOPE_FOUNDRY = "https://ai.azure.com/.default";
export const AZURE_SCOPE_COGNITIVE =
  "https://cognitiveservices.azure.com/.default";

export type AzureApiKind = "chat" | "whisper";

/**
 * 依呼叫的 API 種類選擇 Entra token scope：
 *  - "chat"（v1 路徑 /openai/v1/）→ ai.azure.com 受眾
 *  - "whisper"（傳統 deployments 路徑）→ cognitiveservices 受眾
 */
export function getAzureScopeForApiKind(kind: AzureApiKind): string {
  return kind === "chat" ? AZURE_SCOPE_FOUNDRY : AZURE_SCOPE_COGNITIVE;
}

interface CachedToken {
  accessToken: string;
  expiresAtMs: number;
}

const tokenCache = new Map<string, CachedToken>();
// 提前 60 秒刷新，避免邊界期 token 剛好失效
const EXPIRY_BUFFER_MS = 60_000;

function buildCacheKey(
  credentials: AzureEntraCredentials,
  scope: string,
): string {
  return `${credentials.tenantId}|${credentials.clientId}|${scope}`;
}

interface AzureTokenResult {
  accessToken: string;
  expiresIn: number;
}

export async function getAzureAccessToken(
  credentials: AzureEntraCredentials,
  scope: string,
): Promise<string> {
  if (
    !credentials.tenantId ||
    !credentials.clientId ||
    !credentials.clientSecret
  ) {
    throw new Error("Entra credentials incomplete");
  }

  const cacheKey = buildCacheKey(credentials, scope);
  const now = Date.now();
  const cached = tokenCache.get(cacheKey);
  if (cached && cached.expiresAtMs - EXPIRY_BUFFER_MS > now) {
    return cached.accessToken;
  }

  // 透過 Rust（reqwest）取 token，而非 WebView fetch：WebView 會帶 Origin
  // header，Entra 會以 AADSTS9002326 拒絕瀏覽器來源的 client_credentials
  // token redemption（cross-origin redemption 僅限 SPA client-type）。
  const result = await invoke<AzureTokenResult>("get_azure_entra_token", {
    tenantId: credentials.tenantId,
    clientId: credentials.clientId,
    clientSecret: credentials.clientSecret,
    scope,
  });

  tokenCache.set(cacheKey, {
    accessToken: result.accessToken,
    expiresAtMs: now + result.expiresIn * 1000,
  });

  return result.accessToken;
}

/** 憑證變更或刪除時呼叫，清掉舊 token 快取。 */
export function clearAzureTokenCache(): void {
  tokenCache.clear();
}

File: src-tauri/src/plugins/azure_auth.rs
use serde::{Deserialize, Serialize};

// Entra ID（client credentials）取 bearer token。
//
// 必須在 Rust（reqwest）端發出，而非前端 WebView：WebView 的 fetch 會帶
// Origin header（dev 時為 http://localhost:1420），Entra 會以
// AADSTS9002326（cross-origin token redemption 僅限 SPA client-type）拒絕
// confidential client 的 client_credentials 請求。reqwest 不帶 Origin。

const TOKEN_REQUEST_TIMEOUT_SECS: u64 = 30;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AzureTokenResult {
    pub access_token: String,
    pub expires_in: u64,
}

#[derive(Deserialize)]
struct AadTokenResponse {
    access_token: Option<String>,
    expires_in: Option<u64>,
}

#[derive(Deserialize)]
struct AadError {
    error: Option<String>,
    error_description: Option<String>,
}

#[tauri::command]
pub async fn get_azure_entra_token(
    tenant_id: String,
    client_id: String,
    client_secret: String,
    scope: String,
) -> Result<AzureTokenResult, String> {
    let tenant = tenant_id.trim();
    let client = client_id.trim();
    if tenant.is_empty() || client.is_empty() || client_secret.trim().is_empty() {
        return Err("Entra credentials incomplete".to_string());
    }

    let token_url = format!("https://login.microsoftonline.com/{tenant}/oauth2/v2.0/token");

    let http = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(TOKEN_REQUEST_TIMEOUT_SECS))
        .build()
        .map_err(|e| e.to_string())?;

    let params = [
        ("grant_type", "client_credentials"),
        ("client_id", client),
        ("client_secret", client_secret.as_str()),
        ("scope", scope.as_str()),
    ];

    let response = http
        .post(&token_url)
        .form(&params)
        .send()
        .await
        .map_err(|e| format!("Azure Entra token request failed: {e}"))?;

    let status = response.status();
    let raw = response.text().await.unwrap_or_default();

    if !status.is_success() {
        return Err(format!(
            "Azure Entra token request failed: {}{}",
            status.as_u16(),
            format_aad_detail(&raw)
        ));
    }

    let parsed: AadTokenResponse = serde_json::from_str(&raw)
        .map_err(|e| format!("Failed to parse Entra token response: {e}"))?;
    let access_token = parsed
        .access_token
        .ok_or_else(|| "No access_token in Entra response".to_string())?;

    Ok(AzureTokenResult {
        access_token,
        expires_in: parsed.expires_in.unwrap_or(3600),
    })
}

/// 從 Entra 錯誤回應取出可讀的 AADSTS 說明（不含機密）。
fn format_aad_detail(body: &str) -> String {
    if body.is_empty() {
        return String::new();
    }
    if let Ok(err) = serde_json::from_str::<AadError>(body) {
        if let Some(desc) = err.error_description.or(err.error) {
            let first_line = desc.lines().next().unwrap_or("");
            let snippet: String = first_line.chars().take(300).collect();
            return format!(" — {snippet}");
        }
    }
    let snippet: String = body.chars().take(200).collect();
    format!(" — {snippet}")
}
